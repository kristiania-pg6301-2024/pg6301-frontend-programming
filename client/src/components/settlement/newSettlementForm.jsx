import React, {useState} from 'react';
import {BillDenominationTypes, CoinDenominationTypes} from "./domain";
import {BalanceInputForm} from "./balanceInputForm";

function sumBalance(balance) {
    let sum = 0;
    for (const denonimation in BillDenominationTypes) {
        if (balance[denonimation]?.count) {
            sum += balance[denonimation].count * BillDenominationTypes[denonimation].value;
        }
    }
    for (const denonimation in CoinDenominationTypes) {
        const type = CoinDenominationTypes[denonimation];
        if (balance[denonimation]?.count) {
            sum += balance[denonimation].count * type.value;
        } else if (balance[denonimation]?.weight) {
            sum += Math.round(balance[denonimation].weight / type.gramsPerCoin) * type.value;
        }
    }
    return sum;
}

export function NewSettlementForm() {
    const allDepartments = [
        "Auksjon", "Bilder", "Bøker", "Cafeteria", "Elektro", "Godteri", "Kjøkken", "Møbler", "Popcorn", "Sport",
    ];

    const [settlement, setSettlement] = useState({
        balance: {}
    })
    const sum = sumBalance(settlement.balance);

    return <form>
        <h2>Registrere kontanter</h2>
        <div>
            <div>Avdeling:</div>
            <select onChange={e => setSettlement(old => ({
                ...old,
                department: e.target.value
            }))}>
                <option></option>
                {allDepartments.map(d => <option key={d} value={d}>{d}</option>)}
            </select>
        </div>
        <BalanceInputForm
            balance={settlement.balance}
            setBalance={fn => setSettlement(old => ({
                ...old,
                balance: fn(old.balance)
            }))}
        />
        <div>
            <strong>Sum: {sum}</strong>
        </div>

        <pre>
            {JSON.stringify(settlement, null, 2)}
        </pre>
    </form>;
}
