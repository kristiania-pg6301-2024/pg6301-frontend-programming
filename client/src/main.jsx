import React, { Fragment, useEffect, useState } from "react";
import { createRoot } from "react-dom/client";
import { BrowserRouter, Route, Routes } from "react-router-dom";

const root = createRoot(document.getElementById("root"));
root.render(
  <BrowserRouter>
    <Application />
  </BrowserRouter>,
);

function Application() {
  return (
    <Routes>
      <Route path={"/"} element={<Fragment />} />
      <Route path={"/login/:provider/callback"} element={<LoginCallback />} />
      <Route path={"*"} element={<h1>Not Found</h1>} />
    </Routes>
  );
}

function LoginCallback() {
  return <div>Vennligst vent mens vi fullf√∏rer logginn</div>;
}

function LoginButton() {
  const client_id =
    "34816606807-m5gdfps6ijlopl5psnetie8f3722dmug.apps.googleusercontent.com";
  const discoveryEndpoint =
    "https://accounts.google.com/.well-known/openid-configuration";
  const [authorizationUrl, setAuthorizationUrl] = useState();

  async function createAuthorizationUrl() {
    const configuration = await fetch(discoveryEndpoint);
    const { authorization_endpoint } = await configuration.json();
    const parameters = {
      response_type: "token",
      scope: "profile",
      redirect_uri: window.location.origin + "/login/google/callback",
      client_id,
    };
    setAuthorizationUrl(
      `${authorization_endpoint}?${new URLSearchParams(parameters)}`,
    );
  }

  useEffect(() => {
    createAuthorizationUrl();
  });

  return authorizationUrl ? <a href={authorizationUrl}>Logg inn</a> : null;
}

function FrontPage() {
  const [user, setUser] = useState();
  const [error, setError] = useState();

  async function loadUser() {
    const res = await fetch("/api/userinfo");
    if (res.ok) {
      setUser(await res.json());
    } else {
      setError(`${res.status} ${res.statusText}`);
    }
  }

  useEffect(() => {
    loadUser();
  }, []);

  if (error) {
    return (
      <>
        <h1>Something went wrong</h1>
        <div>{error}</div>
        <LoginButton />
      </>
    );
  }

  if (user) {
    return <h1>You are {user.name}</h1>;
  }

  return <h1>Hello Who Are You</h1>;
}
