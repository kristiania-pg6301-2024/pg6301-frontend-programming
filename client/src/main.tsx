import { createRoot } from "react-dom/client";
import React from "react";
import {
  NewSettlementForm,
  Settlement,
} from "./components/settlements/newSettlementForm";

const root = createRoot(document.getElementById("root")!);

function handleNewSettlement(settlement: Settlement) {
  console.log({ settlement });
}

root.render(<NewSettlementForm onNewSettlement={handleNewSettlement} />);
