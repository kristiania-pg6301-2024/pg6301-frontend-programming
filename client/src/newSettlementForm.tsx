import React from "react";
import { BillTypeValues, CoinTypeValues } from "../../shared/sumBalance";

export function NewSettlementForm() {
  const departments = ["Møbler", "Elektronikk", "Bøker"];

  return (
    <div className={"settlement"}>
      <h1>Register new settlement</h1>
      <div>
        <label>
          Avdeling:
          <br />
          <select>
            {departments.map((department) => (
              <option key={department}>{department}</option>
            ))}
          </select>
        </label>
      </div>
      <div className="money">
        {BillTypeValues.map((bill) => (
          <div>
            <label>
              {bill.label}: <input />
            </label>
          </div>
        ))}
        {CoinTypeValues.map((coin) => (
          <div>
            <label>
              {coin.label}: <input />
            </label>
          </div>
        ))}
      </div>
    </div>
  );
}
