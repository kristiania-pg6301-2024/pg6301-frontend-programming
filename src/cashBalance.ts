export const billTypes = [
  { key: "50kr", value: 50 },
  { key: "100kr", value: 100 },
  { key: "200kr", value: 200 },
  { key: "500kr", value: 500 },
  { key: "1000kr", value: 1000 },
] as const;
type BillDenomination = (typeof billTypes)[number]["key"];

export const coinTypes = [
  { key: "20kr", value: 20, gramsPerCoin: 9.9 },
  { key: "10kr", value: 10, gramsPerCoin: 6.8 },
  { key: "5kr", value: 5, gramsPerCoin: 7.85 },
  { key: "1kr", value: 1, gramsPerCoin: 4.35 },
] as const;
export type CoinType = (typeof coinTypes)[number];
type CoinDenomination = CoinType["key"];

export type CashBalance = Partial<Record<BillDenomination, number>> &
  Partial<Record<CoinDenomination, { count: number } | { grams: number }>>;

export function sumBalance(balance: CashBalance) {
  let sum = 0;
  for (const { key, value } of billTypes) {
    sum += (balance[key] || 0) * value;
  }
  for (const { gramsPerCoin, key, value } of coinTypes) {
    let coin = balance[key];
    if (coin && "count" in coin) {
      sum += coin.count * value;
    } else if (coin) {
      sum += Math.round(coin.grams / gramsPerCoin) * value;
    }
  }
  return sum;
}
