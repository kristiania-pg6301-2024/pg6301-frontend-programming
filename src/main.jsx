import React, {useState} from 'react';
import {createRoot} from "react-dom/client";

const root = createRoot(document.getElementById("root"));

function NewTaskForm({onNewTask}) {

    const [title, setTitle] = useState("");

    function handleSubmit(e) {
        e.preventDefault();
        console.log("Submitting task", { title });
        onNewTask({title})
    }

    return <form onSubmit={handleSubmit}>
        <div>
        <label>
            Title: <input
                type="text"
                value={title}
                onChange={(e) => {
                    setTitle(e.target.value);
                }}
        />
        </label>
        </div>
        <button>Submit new task "{title}"</button>
    </form>;
}

function TaskApplication() {
    const [tasks, setTasks] = useState([])

    function handleNewTask(task) {
        setTasks((prevTasks) => [...prevTasks, task]);
    }

    return <>
        <h1>Here are the tasks we need to do!</h1>

        <ul>
            {tasks.map((task) => (<li>{task.title}</li>))}
        </ul>

        <NewTaskForm onNewTask={handleNewTask} />

        </>;
}

root.render(<TaskApplication />);