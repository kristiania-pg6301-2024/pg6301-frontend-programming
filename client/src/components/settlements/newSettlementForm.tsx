import React, { useState } from "react";
import { billTypes, CashBalance, coinTypes } from "../../../../src/cashBalance";

const departments = ["Furniture", "Books"];

export interface Settlement {
  department: string;
  balance: CashBalance;
}

interface Props {
  onNewSettlement(settlement: Settlement): void;
}

export function NewSettlementForm({ onNewSettlement }: Props) {
  const [department, setDepartment] = useState("");
  const [balance, setBalance] = useState<CashBalance>({});
  function handleSubmit(e: React.FormEvent) {
    e.preventDefault();
    onNewSettlement({ department, balance });
  }

  function hasBalance() {
    return Object.values(balance).some((b) => !!b);
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>New Settlement</h1>

      <div>
        <label htmlFor={"newSettlementDepartment"}>Department: </label>
        <select
          id={"newSettlementDepartment"}
          value={department}
          onChange={(e) => setDepartment(e.target.value)}
        >
          <option></option>
          {departments.map((department) => (
            <option key={department}>{department}</option>
          ))}
        </select>
      </div>

      {billTypes.map(({ key }) => (
        <div key={key}>
          <label htmlFor={`newSettlementBalance_${key}`}>{key}:</label>
          <input
            id={`newSettlementBalance_${key}`}
            type={"number"}
            value={balance[key] || "0"}
            onChange={(e) =>
              setBalance((old) => ({ ...old, [key]: parseInt(e.target.value) }))
            }
          />
        </div>
      ))}
      <div className={"coins"}>
        {coinTypes.map(({ key, gramsPerCoin }) => (
          <div key={key} className={`denomination${key}`}>
            <label>
              {key}:
              <input
                type={"number"}
                className={"count"}
                disabled={balance[key] && "grams" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "count" in balance[key]
                      ? balance[key].count
                      : Math.round(balance[key].grams / gramsPerCoin)
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { count: parseInt(e.target.value) },
                  }))
                }
              />
              <input
                type={"number"}
                className={"grams"}
                disabled={balance[key] && "count" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "grams" in balance[key]
                      ? balance[key].grams
                      : balance[key].count * gramsPerCoin
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { grams: parseFloat(e.target.value) },
                  }))
                }
              />
            </label>
          </div>
        ))}
      </div>
      <button type={"submit"} disabled={!department || !hasBalance()}>
        Submit
      </button>
    </form>
  );
}
