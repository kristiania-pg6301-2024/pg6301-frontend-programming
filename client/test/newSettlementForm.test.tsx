import React from "react";
import { expect, test } from "vitest";
import { NewSettlementForm } from "../src/newSettlementForm";
import { render } from "@testing-library/react";

test("that it hasn't changes", () => {
  const component = render(<NewSettlementForm />);
  expect(
    [...component.baseElement.querySelectorAll(".settlement .money label")].map(
      (l) => l.textContent,
    ),
  ).toEqual([
    "Tusenlapper: ",
    "500kr: ",
    "200kr: ",
    "100kr: ",
    "50kr: ",
    "20kr: ",
    "10kr: ",
    "5kr: ",
    "1kr: ",
  ]);
  expect(component.baseElement).toMatchSnapshot();
});
