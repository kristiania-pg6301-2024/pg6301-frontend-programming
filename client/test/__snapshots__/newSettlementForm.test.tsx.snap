// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`that it hasn't changes 1`] = `
<body>
  <div>
    <div
      class="settlement"
    >
      <h1>
        Register new settlement
      </h1>
      <div>
        <label>
          Avdeling:
          <br />
          <select>
            <option>
              Møbler
            </option>
            <option>
              Elektronikk
            </option>
            <option>
              Bøker
            </option>
          </select>
        </label>
      </div>
      <div
        class="money"
      >
        <div>
          <label>
            Tusenlapper
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            500kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            200kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            100kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            50kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            20kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            10kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            5kr
            : 
            <input />
          </label>
        </div>
        <div>
          <label>
            1kr
            : 
            <input />
          </label>
        </div>
      </div>
    </div>
  </div>
</body>
`;
